python/setup_project/storage_map_widget.py

```diff
@@ -0,0 +1,598 @@
+                self.ui.windows_path_browse.show()
+
+            # set the path values if they've been edited before
+            self.ui.linux_path_edit.setText(edited_linux_path)
+            self.ui.mac_path_edit.setText(edited_mac_path)
+            self.ui.windows_path_edit.setText(edited_windows_path)
+
+        # run this to populate the info label if there are any concerns with
+        # the current input value(s)
+        self.mapping_is_valid()
+
+    def guess_storage(self):
+        """
+        Attempt to guess the most appropriate storage to use for the root.
+        """
+
+        # build lookups by name and id
+        storage_by_id = {}
+        storage_by_name = {}
+
+        # the storages may change as new storages are added manually so we do
+        # this each time this is called
+        storages = self._storage_model.storages
+        for storage in storages:
+            storage_name = storage["code"]
+            storage_id = storage["id"]
+            storage_by_id[storage_id] = storage
+            storage_by_name[storage_name] = storage
+
+        # see if a shotgun storage id is defined in the root info.
+        root_sg_id = self.root_info.get("shotgun_id")
+        if root_sg_id in storage_by_id:
+            # shotgun id defined explicitly for this root. set it!
+            self.local_storage = storage_by_id[root_sg_id]["code"]
+
+        # does name match an existing storage?
+        elif self.root_name in storage_by_name:
+            self.local_storage = self.root_name
+
+        # has this name been mapped before?
+        elif self._best_guess and self._best_guess in storage_by_name:
+            self.local_storage = self._best_guess
+
+        # fall back to requiring the user to manually select an item
+        else:
+            self.local_storage = self._storage_model.CHOOSE_STORAGE_ITEM_TEXT
+
+    def set_count(self, num, total):
+        """Simply sets the text of the count label of the widget. ex (1 of 3)"""
+
+        self.ui.count_lbl.setText("%s of %s" % (num, total))
+
+    def _browse_path(self, line_edit):
+        """Browse and set the path for the supplied edit widget."""
+
+        # create the dialog
+        folder_path = QtGui.QFileDialog.getExistingDirectory(
+            parent=self,
+            caption="Choose Storage Root Folder",
+            options=QtGui.QFileDialog.DontResolveSymlinks |
+                    QtGui.QFileDialog.DontUseNativeDialog |
+                    QtGui.QFileDialog.ShowDirsOnly
+        )
+
+        if folder_path:
+            # folder was browsed, set the path in the widget.
+            line_edit.setText(folder_path)
+
+    def _create_new_storage(self):
+        """Prompt the user for a new storage name."""
+
+        # get all existing storages to provide to the dialog
+        all_storages = self._storage_model.storages
+        existing_storage_names = [storage["code"] for storage in all_storages]
```

Comment:
if possible, could this be all storage names across both active and retired storages?

Reply:
As discussed offline, it's not possible to simply retire a local storage in SG these days. Will leave as is since it's likely an edge case that clients will have old/retired storages.